### PR TITLE
Add AffineLoopInterchange pass

### PR DESCRIPTION
**Pull request by HarshVardhanKumar**
Adds a loop interchange pass to the Affine dialect.
The pass works only for loopnests having rectangular iteration space and no Affine-If statements.
In all other cases, it detects the absence of the above conditions and simply bails out leaving that particular loopnest unchanged.
The pass works for both - imperfectly nested loop nests as well as perfectly nested ones.

### CODE REVIEW

File: mlir/include/mlir/Dialect/Affine/Passes.td [hunk 1]
```diff
@@ -48,6 +48,12 @@ def AffineLoopInvariantCodeMotion
   let constructor = "mlir::createAffineLoopInvariantCodeMotionPass()";
 }
 
+def AffineLoopInterchange
+    : FunctionPass<"affine-loop-interchange">{
+  let summary = "Performs loop interchange of all affine loopnests";
```
Comment (bondhugula) [inline mlir/include/mlir/Dialect/Affine/Passes.td]: loopnests -> loop nests

File: mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp [hunk 2]
```diff
@@ -0,0 +1,1046 @@
+//===- affineLoopInterchange.cpp - Code to perform loop interchange-----===//
```
Comment (bondhugula) [inline mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp]: a -> A
Code -> Pass

File: mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp [hunk 3]
```diff
@@ -0,0 +1,1046 @@
...
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+//
```
Comment (bondhugula) [inline mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp]: Empty lines.

File: mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp [hunk 4]
```diff
@@ -0,0 +1,1046 @@
...
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+#include <algorithm>
+#include <cmath>
+#include <cstdlib>
+#include <iostream>
+#include <set>
+#include <unordered_map>
```
Comment (bondhugula) [inline mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp]: Prune this list to remove unnecessary header list.

File: mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp [hunk 5]
```diff
@@ -0,0 +1,1046 @@
...
+
+#define cache_line_size 64
+
```
Comment (bondhugula) [inline mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp]: Avoid #define's. Instead use const / constexpr variables inside your LoopInterchange struct.

File: mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp [hunk 6]
```diff
@@ -0,0 +1,1046 @@
...
+  std::vector<std::set<Operation *>> referenceGroups;
+};
+} // namespace
+
+/// <summary>
+/// Performs a walk over the loopnest rooted at "forOp" to check if there are
+/// any intermediate "Affine-if" statements in the loopnest.
+/// </summary>
+/// <param name="forOp"></param>
+/// <returns>Returns True if an "Affine-if" statement found.</returns>
```
Comment (bondhugula) [inline mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp]: Just /// comments is enough. The tags aren't needed - please see other code for reference. Include parameter names in backticks.

File: mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp [hunk 7]
```diff
@@ -0,0 +1,1046 @@
...
+bool hasAffineIfStatement(AffineForOp forOp) {
+  int iffound = 0;
+  forOp.walk([&](Operation *op) {
+    if (isa<AffineIfOp>(op))
+      iffound = 1;
+  });
+  return iffound;
```
Comment (bondhugula) [inline mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp]: Use `WalkResult::interrupt()` and `wasInterrupted()` - you won't need the flag. In any case, flags should be `bool` and not `int`.

File: mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp [hunk 8]
```diff
@@ -0,0 +1,1046 @@
...
+}
+
+/// <summary>
+/// Tests if a given AffineForOp loopnest has a rectangular shaped iteration
+/// space.
+/// </summary>
+/// <param name="loopnest"></param>
+/// <returns>Returns True if each AffineForOp in the loopnest has a constant
+/// upper and lower bound. Returns false otherwise.</returns>
+bool isRectangularAffineForLoopNest(
```
Comment (bondhugula) [inline mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp]: `static` method please. You won't need llvm:: on SmallVector.

File: mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp [hunk 9]
```diff
@@ -0,0 +1,1046 @@
...
+  SmallVector<Operation *, 8> loadAndStoreOpInsts;
+  rootForOp.getOperation()->walk([&](Operation *opInst) {
+    if (isa<AffineLoadOp>(opInst) || isa<AffineStoreOp>(opInst)) {
+      loadAndStoreOpInsts.push_back(opInst);
+    }
+  });
+
+  unsigned numOps = loadAndStoreOpInsts.size();
+  for (unsigned i = 0; (i < numOps); ++i) {
+    auto *srcOpInst = loadAndStoreOpInsts[i];
```
Comment (bondhugula) [inline mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp]: Drop `inst` suffixes here. `srcOp`. Also avoid using `auto` unless it's obvious and improves readability - it hampers readability here. Spell out the type. Similarly at other places.

File: mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp [hunk 10]
```diff
@@ -0,0 +1,1046 @@
...
+    uint64_t noSymbols = map.getNumSymbols();
+
+    // #Rows in each accessMatrix: #Dimensions of the MemRef object
+    // #Columns in each accessMatrix: (noDims + noSymbols)
+
+    llvm::SmallVector<std::vector<int64_t>, 0> accessMatrix;
+    std::vector<int64_t> constVector(mapResults.size());
+
+    // accesssing the affine operations.
+    // all for a single instruction
```
Comment (bondhugula) [inline mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp]: All comments have to be proper and grammatically correct sentences. Terminate with a full stop. Fix everywhere else.

File: mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp [hunk 11]
```diff
@@ -0,0 +1,1046 @@
...
+            case AffineExprKind::Constant: {
+              auto constant = rhs.cast<AffineConstantExpr>();
+              Row[position] = constant.getValue();
+              break;
+            }
+            default: {
+            }
+            }
+            break;
+          }
```
Comment (bondhugula) [inline mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp]: ?

File: mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp [hunk 12]
```diff
@@ -0,0 +1,1046 @@
...
+  }
+}
+
+/// <summary>
+/// Checks if an entire column of a 2D matrix is zero
+/// </summary>
+/// <param name="accessMatrix"></param>
+/// <param name="column"></param>
+/// <returns>True if the column is all zeros. False otherwise.</returns>
+bool checkColumnIsZero(llvm::SmallVector<std::vector<int64_t>, 0> &accessMatrix,
```
Comment (bondhugula) [inline mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp]: accessMatrix is read-only. Pass as ArrayRef.

File: mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp [hunk 13]
```diff
@@ -0,0 +1,1046 @@
...
+        DependenceResult result = checkMemrefAccessDependence(
+            srcAccess, dstAccess, depth, &dependenceConstraints, &depComps);
+
+        if (mlir::hasDependence(result)) {
+          llvm::SmallVector<int64_t, 8> components;
+
+          // averaging the upper and lower bounds on dependenceComponents
+          for (auto depComp : depComps) {
+            if (depComp.lb.getValue() * depComp.ub.getValue() >=
+                0) { // both upper and lower bound have same signs
```
Comment (bondhugula) [inline mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp]: No trailing comment.

**Review by bondhugula — CHANGES_REQUESTED** [review]
Please see this initial set of comments - the same fixes need to be made throughout.

**Review by bondhugula — CHANGES_REQUESTED** [review]
Please add test cases without which this would be incomplete.

**Comment by HarshVardhanKumar** [comment]
Will do the modifications.

**Comment by bondhugula** [comment]
Please don't close and create a new PR. Instead, just update and force push.

**Pull request closed without merging**